avocado/utils/software_manager/main.py

```diff
@@ -32,18 +34,21 @@ def main():
             log.info("Packages %s installed successfully", args)
         else:
             log.error("Failed to install %s", args)
+            exitcode |= exit_codes.UTILITY_MEDIUM
```

Comment:
I have to confess that I have mix feelings about these "generic" returns. This sounds to me, something similar to having an error alert with message like "a catastrophic error happen!", but without much details.

I understand that we have the log, but IMO we should try to express this into return codes as well and try to follow some basic convention.

Is there any reason for not using the default `/usr/include/sysexits.h` codes?

Reply:
I don't see too much overlapping of the definitions in `/usr/include/sysexits.h` and the ones needed here.  But I think you have about about the "generic" exit codes.  I'd avoid trying to define generic codes not needed right now.

Reply:
> I have to confess that I have mixed feelings about these "generic" returns. This sounds to me, something similar to having an error alert with a message like "a catastrophic error happen!", but without many details.
>

This is a command-line utility; the specific message is already displayed to the user in the previous line. The intent of the exit code is to be used by other programs, like when checking the execution of, for example, this, https://github.com/avocado-framework/avocado/pull/4275/files#diff-ea6bb6adabe1ef27fa0285187eff623b7ccca21a35523f67bf62eff194cff48bR105

Right now, when a program, like the Requirements Resolver, runs the `avocado-software-manager`, it does not know if the command succeeded or not without having to parse the output of the command looking for the exit code.
 
> I understand that we have the log, but IMO we should try to express this into return codes as well and try to follow some basic convention.
> 
> Is there any reason for not using the well know `/usr/include/sysexits.h` codes?

I don't think so, we can use it. The intent here is to create an easy reference to users developing a new command-line utility for Avocado. Instead of letting the users choose the codes from a big list, we would suggest using one of the codes from the `exit_codes.py` file.

Reply:
@willianrampazzo As I said, I understand that we have the log line, but for the practical aspect, IMO it is better to have meaningful exit codes, so I don't need to check for log messages. Logs can be disabled for many reasons. Exit codes are not only for software but also for users/debugging purposes. The way is here in this patch, is something like: "A generic error", "a more generic error", and a "less generic error".

For me, the definitions of what is LOW, MEDIUM, and HIGH are confusing and could be interpreted depending on the reader. Also, we would be better if we don't require that our users read a new document or a source code file to undertand our exit codes.

The idea is nobble, and I'm in favor for it. Unless I'm missing something, I would prefer to redirect developers to a common Unix standard instead of creating a new one.

Reply:
Okay, I got your point now, it is more related to the variable meaning. I was thinking of something else. Well, that works for me too. I just need exit codes here to move with the Requirements Resolver :D

Reply:
Humm, reading the `/usr/include/sysexits.h` file, I got really confused about which code I would use for a check failure. In this case, I return my previous statement and think a future reference for users inside the utility directory will make things easy.

Here is the list inside `sysexits` as a quick reference:

```
#define EX_OK		0	/* successful termination */

#define EX__BASE	64	/* base value for error messages */

#define EX_USAGE	64	/* command line usage error */
#define EX_DATAERR	65	/* data format error */
#define EX_NOINPUT	66	/* cannot open input */
#define EX_NOUSER	67	/* addressee unknown */
#define EX_NOHOST	68	/* host name unknown */
#define EX_UNAVAILABLE	69	/* service unavailable */
#define EX_SOFTWARE	70	/* internal software error */
#define EX_OSERR	71	/* system error (e.g., can't fork) */
#define EX_OSFILE	72	/* critical OS file missing */
#define EX_CANTCREAT	73	/* can't create (user) output file */
#define EX_IOERR	74	/* input/output error */
#define EX_TEMPFAIL	75	/* temp failure; user is invited to retry */
#define EX_PROTOCOL	76	/* remote error in protocol */
#define EX_NOPERM	77	/* permission denied */
#define EX_CONFIG	78	/* configuration error */

#define EX__MAX	78	/* maximum listed value */
```

Reply:
@willianrampazzo, I'm trying to follow the @clebergnu suggestion:

> "Anyone, the main point remains, in the sense that "success" should be 0, failures >0, and crashes == -1."

The question here is: how we are going to categorize our failures? Or we use a simple approach (0,1,-1), or we need to categorize our failures (>1 AND <64). My previous comment was more about avoiding too much generic categories.

I still think that we could use those categories, but if we are not willing to adopt -1, or even `EX_SOFTWARE` for your example, looks like you answered my initial question, why we should not use it as it is.

At this point, my concerns are:

 a) removing from the developers the ability to use the well know codes; For instance: if someone is developing a tool, that can "crash" because of a permission denied or an I/O error, IMO would be important to give that developer the option to use the very well know exit codes in the UNIX world.

 b) Creating custom categories will increase the learning curve of our users/developers; I would use this as last shot.

 c) Using multiple categories that are all generics. I'm ok with having a "general exception" for errors that don't fit in the current categories, but not sure (yet) if having "HIGH', "LOW", "MEDIUM" generic errors are fine.

Reply:
> @willianrampazzo, I'm trying to follow the @clebergnu suggestion:
> 
> > "Anyone, the main point remains, in the sense that "success" should be 0, failures >0, and crashes == -1."
> 
> The question here is: how we are going to categorize our failures? Or we use a simple approach (0,1,-1), or we need to categorize our failures (>1 AND <64). My previous comment was more about avoiding too much generic categories.
> 
> I still think that we could use those categories, but if we are not willing to adopt -1, or even `EX_SOFTWARE` for your example, looks like you answered my initial question, why we should not use it as it is.
> 
> At this point, my concerns are:
> 
> a) removing from the developers the ability to use the well know codes; For instance: if someone is developing a tool, that can "crash" because of a permission denied or an I/O error, IMO would be important to give that developer the option to use the very well know exit codes in the UNIX world.

Agreed, we should not limit the code options for the developers to those defined in the `exit_codes.py`.

> 
> b) Creating custom categories will increase the learning curve of our users/developers; I would use this as last shot.
> 

Right now I'll limit it to the suggestions from @clebergnu, 0, 1 and -1. That should be enough for now.

> c) Using multiple categories that are all generics. I'm ok with having a "general exception" for errors that don't fit in the current categories, but not sure (yet) if having "HIGH', "LOW", "MEDIUM" generic errors are fine.

That is why I mentioned in the PR description that I was lacking imagination for the names :D I also do not agree those are good names. I was expecting some suggestions in the review.